csharp/src/Apache.Arrow/Ipc/ArrowStreamWriter.cs

```diff
@@ -272,7 +272,7 @@ public Task WriteBufferAsync(ArrowBuffer arrowBuffer, CancellationToken cancella
                 var span = arrowBuffer.Span;
                 buffer = ArrayPool<byte>.Shared.Rent(span.Length);
                 span.CopyTo(buffer);
-                return BaseStream.WriteAsync(buffer, 0, buffer.Length, cancellationToken);
+                return BaseStream.WriteAsync(buffer, 0, span.Length, cancellationToken);
```

Comment:
@eerhardt, not directly related to your changes, but I've been trying them out, and one thing that's related to minimizing allocations that I bumped up against is that ArrowStreamWriter always disposes the underlying stream.  That's very undesirable if the arrow data being written is just part of a larger chunk of data being written out, such as if there's a header and footer that need to wrap it.  ArrowStreamWriter should have a ctor that allows for leaving the underlying stream open upon dispose, as exists for built-in .NET types like StreamWriter, Deflatestream, etc.

Reply:
I opened https://github.com/apache/arrow/pull/3782 for this.  I can fix it up after your changes to merge better, e.g. for tests.